File: day01.py
"""Solution to day 1 of Advent of Code"""

from get_input import get_input, line_parser


def part1(lines, offset=1):
    return sum(1 for a, b in zip(lines, lines[offset:]) if a < b)


def part2(lines):
    return part1(lines, offset=3)


if __name__ == "__main__":
    LINES = line_parser(get_input(day=1, year=2021))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day01.rs
use common::get_input;

type Input = Vec<usize>;

fn part1(input: &Input) -> usize {
    count_increases(input, 1)
}

fn part2(input: &Input) -> usize {
    count_increases(input, 3)
}

fn count_increases(items: &Vec<usize>, offset: usize) -> usize {
    items.iter().zip(items.iter().skip(offset))
        .filter(|(a, b)| a < b)
        .count()
}

fn parse(text: String) -> Input {
    text.trim()
        .split('\n')
        .map(|n| n.parse::<usize>().unwrap())
        .collect()
}

fn main() {
    let input = parse(get_input(01, 2021));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
